### PR TITLE
Red inclusion

### LINKED ISSUE #1029: Include transactions from red blocks and bestow the rewards and fees to merging blocks
We decided that txns from red blocks should also be included.

Note: The rewards (fees + CB txn) will go to the merging block, rather than the actual block, more on that should be detailed in the monetary policy issue.

Recall that the merge set of a block is the set of all blocks in its past which are not in the past of its selected parent. Formally, `B.MergeSet = B.Past \ B.SelectedParent.Past`. When calculating the UTXO set of a block B, all fees and block rewards of all red blocks in B.MergeSet should go to B.

Note that the implication of the change is that, from the point of view of virtual, the rewards of any red block will belong to the earliest chain block in its future.

Motivation: We have decided to include red blocks (cf. ). However, there is currently no incentive for miners to do so, and in fact any form of DAA which regulates block rate (rather than regulating blue block rate, which is inherently broken) inherently decentivizes miners from pointing at red blocks (as these will increase the difficulty without increasing the score). Bestowing the rewards to the merger incentivizes merging red blocks.

### PR DESCRIPTION
**Pull request by someone235**
Closes #1029

### CODE REVIEW

**Comment by elichai** [comment]
Does this include https://github.com/kaspanet/kaspad/issues/1259 or is it separate?

**Comment by someone235** [comment]
> Does this include #1259 or is it separate?

Yeah sorry for not mentioning it. For now red block reward and fees are dismissed. I opened an issue to fix it #1276

File: app/protocol/flows/blockrelay/ibd.go [hunk 1]
```diff
@@ -91,7 +91,7 @@ func (flow *handleRelayInvsFlow) runIBDIfNotRunning(highHash *externalapi.Domain
 	return nil
 }
 
-func (flow *handleRelayInvsFlow) fetchUTXOSetIfMissing() (bool, error) {
+func (flow *handleRelayInvsFlow) fetchUTXOSetIfMissing() (bool, error) { // TODO: Remove this
```
Comment (stasatdaglabs) [inline app/protocol/flows/blockrelay/ibd.go:94]: By chance, did you mean to remove this?

File: go.mod [hunk 2]
```diff
@@ -6,6 +6,7 @@ require (
 	github.com/btcsuite/go-socks v0.0.0-20170105172521-4720035b7bfd
 	github.com/btcsuite/winsvc v1.0.0
 	github.com/davecgh/go-spew v1.1.1
+	github.com/go-test/deep v1.0.7
```
Comment (stasatdaglabs) [inline go.mod:9]: Please remove this dependency and run `go mod tidy`

File: domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go [hunk 3]
```diff
@@ -151,6 +154,215 @@ func TestDoubleSpends(t *testing.T) {
 	})
 }
 
+// TestTransactionAcceptance checks that blue blocks transactions are favoured above
+// red blocks transactions, and that the block reward is payed only for blue blocks.
```
Comment (stasatdaglabs) [inline domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go:158]: Typo: payed -> paid

File: domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go [hunk 4]
```diff
@@ -151,6 +154,215 @@ func TestDoubleSpends(t *testing.T) {
...
+func TestTransactionAcceptance(t *testing.T) {
+	testutils.ForAllNets(t, true, func(t *testing.T, params *dagconfig.Params) {
+		params.BlockCoinbaseMaturity = 0
+
+		factory := consensus.NewFactory()
+		tc, teardown, err := factory.NewTestConsensus(params, "TestTransactionAcceptance")
```
Comment (stasatdaglabs) [inline domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go:164]: Please don't use two-letter abbreviations.

File: domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go [hunk 5]
```diff
@@ -151,6 +154,215 @@ func TestDoubleSpends(t *testing.T) {
...
+		if err != nil {
+			t.Fatalf("Error creating fundingBlock1: %+v", err)
+		}
+
+		fundingBlock2Hash, _, err := tc.AddBlock([]*externalapi.DomainHash{fundingBlock1Hash}, nil, nil)
+		if err != nil {
+			t.Fatalf("Error creating fundingBlock2: %+v", err)
+		}
+
+		// Generate fundingBlock3 to pay for fundingBlock1 and fundingBlock2
```
Comment (stasatdaglabs) [inline domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go:180]: fundingBlock3 funds fundingBlock2 only. Please correct this comment.

**Review by stasatdaglabs — CHANGES_REQUESTED** [review]

**Review by stasatdaglabs — APPROVED** [review]

**Pull request merged**